firedrake/preconditioners/asm.py

```diff
@@ -267,8 +268,122 @@ def get_patches(self, V):
                 if dof <= 0:
                     continue
                 off = section.getOffset(p)
-                indices = numpy.arange(off*V.value_size, V.value_size * (off + dof), dtype='int32')
+                indices = numpy.arange(off*V.value_size, V.value_size * (off + dof), dtype=IntType)
                 iset = PETSc.IS().createGeneral(indices, comm=COMM_SELF)
                 ises.append(iset)
 
         return ises
+
+
+def get_basemesh_nodes(W):
+    pstart, pend = W.mesh().topology_dm.getChart()
+    section = W.dm.getDefaultSection()
+    # location of first dof on an entity
+    basemeshoff = numpy.empty(pend - pstart, dtype=IntType)
+    # number of dofs on this entity
+    basemeshdof = numpy.empty(pend - pstart, dtype=IntType)
+    # number of dofs stacked on this entity in each cell
+    basemeshlayeroffset = numpy.empty(pend - pstart, dtype=IntType)
+
+    # For every base mesh entity, what's the layer offset?
+    layer_offsets = numpy.full(W.node_set.total_size, -1, dtype=IntType)
+    layer_offsets[W.cell_node_map().values_with_halo] = W.cell_node_map().offset
+    nlayers = W.mesh().layers
+
+    for p in range(pstart, pend):
+        dof = section.getDof(p)
+        off = section.getOffset(p)
+        if dof == 0:
+            dof_per_layer = 0
+            layer_offset = 0
+        else:
+            layer_offset = layer_offsets[off]
+            assert layer_offset >= 0
+            dof_per_layer = dof - (nlayers - 1) * layer_offset
+
+        basemeshoff[p - pstart] = off
+        basemeshdof[p - pstart] = dof_per_layer
+        basemeshlayeroffset[p - pstart] = layer_offset
+
+    return basemeshoff, basemeshdof, basemeshlayeroffset
+
+
+class ASMHexStarPC(ASMPatchPC):
```

Comment:
So `ASMStarPC` doesn't do the right thing on extruded meshes.

Should `ASMStarPC` raise a warning on extruded meshes?

Does `ASMHexStarPC` work on extruded meshes other than hexes? (Prisms, for example?)

Reply:
`ASMHexStar` works well for prisms (when I implement Pavarino I get h- and p- robust iteration counts). So maybe it sould be renamed?

Reply:
Yes, I think it should be renamed. `ASMExtrudedStarPC`?